### PR TITLE
Handle & and | types when computing tuple arity

### LINKED ISSUE #15302: inlined match on tuple, and expanding with `*:` causes class cast exception at runtime
## Compiler version

3.1.3-RC4, 3.2.0-RC1-bin-20220526-1ea177d-NIGHTLY-git-1ea177d

## Minimized code

```Scala
inline def flatConcat2[A, B](a: A, b: B) =
  b match
    case bs: *:[bh, bt]  => a *: bs

val x: (Int, Int, Int) = flatConcat2(1, (2,3))

```

## Output

`java.lang.ClassCastException: class scala.Tuple3 cannot be cast to class scala.Tuple1 (scala.Tuple3 and scala.Tuple1 are in unnamed module of loader sbt.internal.ScalaLibraryClassLoader @3401c1d7)`
pointing to the end of the flatConcat2 definition
	

## Expectation

This works if the `inline` is removed
Should not throw a ClassCastException.

Note, the match type of the def is:
```
type FlatConcat2[A, B] = B match
  case b *: bs => A *: b *: bs
```
but adding that does not change anything

### PR DESCRIPTION
**Pull request by nicolasstucki**
Fixes #15302

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TypeUtils.scala [hunk 1]
```diff
@@ -58,10 +58,15 @@ object TypeUtils {
         if (arity < 0) arity else arity + 1
       case self: SingletonType =>
         if self.termSymbol == defn.EmptyTupleModule then 0 else -1
-      case self if defn.isTupleClass(self.classSymbol) =>
-        self.dealias.argInfos.length
+      case AndType(tp1, tp2) =>
+        tp1.tupleArity.max(tp2.tupleArity)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/transform/TypeUtils.scala:62]: should this check that both sides are equal? `(Int, Int) & (Int, Int, Int)` doesn't make sense to me

**Comment by nicolasstucki** [comment]
After rebasing, I had some conflicts with #15250. The second commit ([e8e30ed](https://github.com/lampepfl/dotty/pull/15330/commits/e8e30ed1a27a5ebfc161a14cffafabfea5e9f8e5)) fixes those conflicts.

File: compiler/src/dotty/tools/dotc/transform/TypeUtils.scala [hunk 2]
```diff
@@ -51,41 +51,44 @@ object TypeUtils {
 
     /** The arity of this tuple type, which can be made up of EmptyTuple, TupleX and `*:` pairs,
      *  or -1 if this is not a tuple type.
-     *
-     *  @param relaxEmptyTuple if true then TypeRef(EmptyTuple$) =:= EmptyTuple.type
      */
-    def tupleArity(relaxEmptyTuple: Boolean = false)(using Context): Int = self match {
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/transform/TypeUtils.scala:57]: why was this parameter removed? will erasure remain the same for types where `EmptyTuple.type` was widened to `TypeRef(EmptyTuple$)`?

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/TypeUtils.scala:57]: ok it seems to work for refined printer I see you made the change, disregard all :)

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/TypeUtils.scala:57]: For other that may read this thread: the important thing here is that `tupleElementTypes` handles the `EmptyTuple` type alias but  `tupleArity` does not (see comment on dealiasing)

**Review by bishabosha — APPROVED** [review]

**Pull request merged**